src/Common/IFileCachePriority.h

```diff
@@ -0,0 +1,106 @@
+
+/// IFileCachePriority is used to maintain the priority of cached data.
+class IFileCachePriority
+{
+public:
+    class IIterator;
+    friend class IIterator;
+
+    using ReadIterator = std::shared_ptr<const IIterator>;
+    using WriteIterator = std::shared_ptr<IIterator>;
+
+    friend class FileCache;
+    using Key = FileCacheKey;
+
+    struct FileCacheRecord
+    {
+        Key key;
+        size_t offset;
+        size_t size;
+        size_t hits = 0;
+
+        FileCacheRecord(const Key & key_, size_t offset_, size_t size_) : key(key_), offset(offset_), size(size_) { }
+    };
+
+    /// It provides an iterator to traverse the cache priority. Under normal circumstances,
+    /// the iterator can only return the records that have been directly swapped out.
+    /// For example, in the LRU algorithm, it can traverse all records, but in the LRU-K, it
+    /// can only traverse the records in the low priority queue.
+    class IIterator
+    {
+    public:
+        virtual ~IIterator() = default;
+
+        virtual Key key() const = 0;
+
+        virtual size_t offset() const = 0;
+
+        virtual size_t size() const = 0;
+
+        virtual size_t hits() const = 0;
+
+        /// Point the iterator to the next higher priority cache record.
+        virtual void next() const = 0;
+
+        virtual bool valid() const = 0;
+
+        /// Mark a cache record as recently used, it will update the priority
+        /// of the cache record according to different cache algorithms.
+        virtual void use(std::lock_guard<std::mutex> &) = 0;
+
+        /// Deletes an existing cached record.
+        virtual void remove(std::lock_guard<std::mutex> &) = 0;
+
+        /// Get an iterator to handle write operations. Write iterators should only
+        /// be allowed to call remove, use and incrementSize methods.
+        virtual WriteIterator getWriteIterator() const = 0;
+
+        virtual void incrementSize(size_t, std::lock_guard<std::mutex> &) = 0;
+
+        /// Repoint the iterator to the record with the lowest priority.
+        virtual void seekToLowestPriority() const = 0;
+    };
+
+public:
+    virtual ~IFileCachePriority() = default;
+
+    /// Add a cache record that did not exist before, and throw a
+    /// logical exception if the cache block already exists.
+    virtual WriteIterator add(const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock) = 0;
+
+    /// Query whether a cache record exists. If it exists, return true. If not, return false.
+    virtual bool contains(const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock) = 0;
+
+    virtual void removeAll(std::lock_guard<std::mutex> & cache_lock) = 0;
+
+    /// Returns an iterator pointing to the lowest priority cached record.
+    /// We can traverse all cached records through the iterator's next().
+    virtual ReadIterator getNewIterator(std::lock_guard<std::mutex> & cache_lock) = 0;
```

Comment:
May be worth to name it `getLowestPriorityIterator`?
Though then it would not be obvious why it returns ReadIterator. 
May be do like this ?
```
/// Get Iterator for read-only operations
virtual const Iterator getLowestPriorityIterator(std::lock_guard<std::mutex> & cache_lock) = 0;

/// Get iterator for read-write operations
virtual Iterator getLowestPriorityIterator(std::lock_guard<std::mutex> & cache_lock) = 0;
```
what do you think?